src/main/java/net/mcreator/blockly/feature/BlocklyToFeature.java

```diff
@@ -19,86 +19,63 @@
-	 * @param firstBlockName     <p>The name of the start block</p>
-	 * @param sourceXML          <p>The XML code used by Blockly</p>
-	 * @param templateGenerator  <p>The folder location in each {@link Generator} containing the code template files<p>
-	 */
-	public BlocklyToFeature(Workspace workspace, String firstBlockName, String sourceXML,
-			TemplateGenerator templateGenerator, IBlockGenerator... externalGenerators)
-			throws TemplateGeneratorException {
-		super(workspace, templateGenerator, externalGenerators);
+	public BlocklyToFeature(Workspace workspace, String sourceXML, TemplateGenerator templateGenerator,
+			IBlockGenerator... externalGenerators) throws TemplateGeneratorException {
+		super(workspace, BlocklyEditorType.FEATURE.getStartBlockName(), sourceXML, templateGenerator,
+				externalGenerators);
+	}
+
+	@Override public void preInitialization() {
+		super.preInitialization();
 		featureConfigurationCode = new StringBuilder();
 		featureType = "";
+	}
 
-		blockGenerators.add(new MCItemBlock());
+	@Override public void preBlocksPlacement(Document doc, Element startBlock) {
+		super.preBlocksPlacement(doc, startBlock);
 
-		if (sourceXML != null) {
+		// Add the feature to the feature code
+		Element feature = XMLUtil.getFirstChildrenWithName(startBlock, "value");
+		if (feature != null) {
 			try {
-				final Document doc = DocumentBuilderFactory.newInstance().newDocumentBuilder()
-						.parse(new InputSource(new StringReader(sourceXML)));
-				doc.getDocumentElement().normalize();
-
-				Element start_block = BlocklyBlockUtil.getStartBlock(doc, firstBlockName);
-
-				// if there is no start block, we return empty string
-				if (start_block == null)
-					throw new ParseException("Could not find start block!", -1);
-
-				// Add the feature to the feature code
-				Element feature = XMLUtil.getFirstChildrenWithName(start_block, "value");
-				if (feature != null) {
-					featureConfigurationCode.append(directProcessOutputBlock(this, feature));
-					Element featureBlock = XMLUtil.getFirstChildrenWithName(feature, "block");
-					if (featureBlock != null)
-						this.featureType = featureBlock.getAttribute("type");
-				}
-				else
-					addCompileNote(new BlocklyCompileNote(BlocklyCompileNote.Type.ERROR,
-							L10N.t("blockly.errors.features.missing_feature")));
-
-				List<Element> base_blocks = BlocklyBlockUtil.getBlockProcedureStartingWithNext(start_block);
-				if (base_blocks.isEmpty())
-					addCompileNote(new BlocklyCompileNote(BlocklyCompileNote.Type.ERROR,
-							L10N.t("blockly.errors.features.missing_placement")));
-				processBlockProcedure(base_blocks);
+				featureConfigurationCode.append(directProcessOutputBlock(this, feature));
+				Element featureBlock = XMLUtil.getFirstChildrenWithName(feature, "block");
+				if (featureBlock != null)
+					this.featureType = featureBlock.getAttribute("type");
 			} catch (TemplateGeneratorException e) {

```

Comment:
I wasn't sure what to do for this part because to rethrow it, the general method has to throw the exception.

Reply:
You could add `throws TemplateGeneratorException` to the method in `BlocklyToJava`. The override in `BlocklyToProcedure` doesn't need to be changed, since it doesn't throw anything